prov/efa/src/rxr/rxr_ep.c

```diff
@@ -1803,60 +1804,72 @@ static inline void rxr_ep_check_peer_backoff_timer(struct rxr_ep *ep)
  * @param[in]	ep		end point
  * @param[in]	cqe_to_process	max number of cq entry to poll and process
  */
-static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
-				      size_t cqe_to_process)
+static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep)
 {
-	struct ibv_wc ibv_wc;
+	bool should_end_poll = false;
 	struct efa_cq *efa_cq;
+	/* Initialize an empty ibv_poll_cq_attr struct for ibv_start_poll.
+	 * EFA expects .comp_mask = 0, or otherwise returns EINVAL.
+	 */
+	struct ibv_poll_cq_attr poll_cq_attr = {.comp_mask = 0};
 	struct efa_av *efa_av;
 	struct efa_ep *efa_ep;
 	struct rxr_pkt_entry *pkt_entry;
 	ssize_t ret;
-	int i, err, prov_errno;
+	int err, prov_errno;
 
 	efa_ep = container_of(ep->rdm_ep, struct efa_ep, util_ep.ep_fid);
 	efa_av = efa_ep->av;
 	efa_cq = container_of(ep->rdm_cq, struct efa_cq, util_cq.cq_fid);
-	for (i = 0; i < cqe_to_process; i++) {
-		ret = ibv_poll_cq(efa_cq->ibv_cq, 1, &ibv_wc);
-
-		if (ret == 0)
-			return;
 
-		if (OFI_UNLIKELY(ret < 0 || ibv_wc.status)) {
-			if (ret < 0) {
-				efa_eq_write_error(&ep->util_ep, -ret, -ret);
-				return;
-			}
+	/* Call ibv_start_poll only once */
+	ret = ibv_start_poll(efa_cq->ibv_cq_ex, &poll_cq_attr);
+	if (!ret)
+		should_end_poll = true;
+
+	do {
+		if (OFI_UNLIKELY(ret != 0)) {
+		    if (ret == ENOENT) {
+				/* When no completions are available on the CQ, ENOENT is returned,
+				 * but the CQ remains in a valid state.
+				 */
+				break;
+		    }
+			/* ret = 0 iff start_poll or next_poll was successful */
+			if (ret < 0)
+				ret = -ret;
+			efa_eq_write_error(&ep->util_ep, ret, ret);
+			break;
+		}
 
-			pkt_entry = (void *)(uintptr_t)ibv_wc.wr_id;
-			err = ibv_wc.status;
-			prov_errno = ibv_wc.status;
-			if (ibv_wc.opcode == IBV_WC_SEND) {
+		if (ret || efa_cq->ibv_cq_ex->status) {
+			pkt_entry = (void *)(uintptr_t)efa_cq->ibv_cq_ex->wr_id;
+			err = efa_cq->ibv_cq_ex->status;
+			prov_errno = efa_cq->ibv_cq_ex->status;
+			if (ibv_wc_read_opcode(efa_cq->ibv_cq_ex) == IBV_WC_SEND) {
 #if ENABLE_DEBUG
 				ep->failed_send_comps++;
 #endif
 				rxr_pkt_handle_send_error(ep, pkt_entry, err, prov_errno);
 			} else {
-				assert(ibv_wc.opcode == IBV_WC_RECV);
+				assert(ibv_wc_read_opcode(efa_cq->ibv_cq_ex) == IBV_WC_RECV);
 				rxr_pkt_handle_recv_error(ep, pkt_entry, err, prov_errno);
 			}
-
-			return;
+			break;
 		}
 
-		pkt_entry = (void *)(uintptr_t)ibv_wc.wr_id;
+		pkt_entry = (void *)(uintptr_t)efa_cq->ibv_cq_ex->wr_id;
 
-		switch (ibv_wc.opcode) {
+		switch (ibv_wc_read_opcode(efa_cq->ibv_cq_ex)) {
```

Comment:
`opcode` has been use multiple times, maybe store it in a variable?

Reply:
I did that and realized that it's only used once on the happy path, so I did not allocate a new variable.